File: Typescript/Design Patterns/Behavioral/Visitor.ts
interface IVisitor {
  visitBook(book: Book): void
  visitElectronics(electronics: Electronics): void
}

interface IProduct {
  accept(visitor: IVisitor): void
}

class Book implements IProduct {
  constructor(public price: number) {}

  accept(visitor: IVisitor) {
    visitor.visitBook(this)
  }
}
class Electronics implements IProduct {
  constructor(public price: number) {}

  accept(visitor: IVisitor) {
    visitor.visitElectronics(this)
  }
}

class TaxVisitor implements IVisitor {
  visitBook(book: Book) {
    const tax = book.price * 0.05
    console.log(`Book tax: $${tax}`)
  }

  visitElectronics(e: Electronics) {
    const tax = e.price * 0.18
    console.log(`Electronics tax: $${tax}`)
  }
}

const products: IProduct[] = [
  new Book(30),
  new Electronics(500)
]

const taxVisitor = new TaxVisitor()

for (const product of products) {
  product.accept(taxVisitor)
}

File: C#/Design Patterns/Behavioral/Visitor.cs
using System;
using System.Collections.Generic;
interface IVisitor
{
  void VisitBook(Book book);
  void VisitElectronics(Electronics electronics);
}
interface IProduct
{
  void Accept(IVisitor visitor);
}
class Book : IProduct
{
  public int price;
  public Book(int price) { this.price = price; }

  public void Accept(IVisitor visitor) { visitor.VisitBook(this); }
}
class Electronics : IProduct
{
  public int price;
  public Electronics(int price)
  { this.price = price; }

  public void Accept(IVisitor visitor) { visitor.VisitElectronics(this); }
}

class TaxVisitor : IVisitor
{
  public void VisitBook(Book book)
  {
    var tax = book.price * 0.05;
    Console.WriteLine($"Book tax: ${tax}");
  }

  public void VisitElectronics(Electronics e)
  {
    var tax = e.price * 0.18;
    Console.WriteLine($"Electronics tax: ${tax}");
  }
}

class Program
{
  public static void Main(string[] args)
  {
    List<IProduct> products = new List<IProduct>
        {
            new Book(30),
            new Electronics(500)
        };

    TaxVisitor taxVisitor = new TaxVisitor();
    products.ForEach(product => product.Accept(taxVisitor));
  }
}